mlir/lib/Conversion/TosaToRock/TosaToRock.cpp

```diff
@@ -1448,6 +1466,67 @@ struct AttentionRewritePattern : public OpRewritePattern<tosa::MatMulOp> {
     return failure();
   }
 
+  /**
+   * Attempts to match and extract a Log-Sum-Exp (LSE) pattern from TOSA
+   * operations.
+   *
+   * This function traverses the users of a reduce sum operation to identify a
+   * complete LSE computation pattern, which typically consists of:
+   * 1. A reduce_max operation to find the maximum values (in some cases this
+   * might not exist)
+   * 2. Subtraction of the max from original values (implicit in the pattern)
+   * 3. Exponential and sum operations (represented by reduceSum)
+   * 4. A logarithm operation on the result
+   * 5. Addition of the original max values back
+   *
+   * Note that reduceSum and reduceMax are given.
+   *
+   * The LSE pattern: log(sum(exp(x - max(x)))) + max(x)
+   */
+  Value getLSE(Operation *reduceSum, Operation *reduceMax,
+               tosa::LogOp logOp = nullptr) const {
+    for (auto *user : reduceSum->getUsers()) {
+      if (auto op = dyn_cast<tosa::LogOp>(user)) {
+        // we already found a log
+        if (logOp != nullptr)
+          return nullptr;
+        Value val = getLSE(op, reduceMax, op);
+        if (val)
+          return val;
+      } else if (auto addOp = dyn_cast<tosa::AddOp>(user)) {
+        auto logOpFromAdd =
+            getDefiningNonReshapeOp<tosa::LogOp>(addOp.getInput1());
+        if (!logOpFromAdd)
+          logOpFromAdd =
+              getDefiningNonReshapeOp<tosa::LogOp>(addOp.getInput2());
+
+        // must match the logOp
+        if (!logOp || logOp != logOpFromAdd)
+          return nullptr;
```

Comment:
This assumes `getUsers()` list first hits `logOp` and then `addOp`. Is that ordering guranteed ?